### PR TITLE
fix(forward): detach tensors from computationnal graph to release gpu…

### PR DESCRIPTION
**Pull request by nicolas-gaude**
detach tensors from computationnal graph to release gpu memory during forward pass.
this limits the needs for GPU to maintain batched embeddings tensors output of shape = [ batch_size, 768]

### CODE REVIEW

**Comment by nreimers** [comment]
Thanks for the PR.

Is there a reason why you do:
```
embeddings = np.copy(embeddings.cpu().detach().numpy())

```
instead of:
```
embeddings = embeddings.cpu().detach().numpy()
```
?

Best
Nils

**Comment by nicolas-gaude** [comment]
Hi,

in fact the np.copy() operation allows garbage python collector to release the underlying gpu torch tensor.

to prove this, let's put this debugging snippet code into the batch loop ( `for batch_idx in iterator:` ) of the `SentenceTransformer.encode()` function 

```
            import gc
            sizes = []
            gc.collect()
            for obj in gc.get_objects():
                try:
                    if torch.is_tensor(obj) or (hasattr(obj, 'data') and torch.is_tensor(obj.data)):
                        sizes.append(obj.size())
                except:
                    pass
            from collections import Counter
            print('allocated tensors summary :', Counter(sizes))
            print('allocated tensors count', len(sizes))
            print('-----------------------')
```
(this snippet code prints every tensors allocated along the loop iterations.)

with` embeddings = embeddings.cpu().detach().numpy()` we observe the following trace during embeddings :-1: 
```
allocated tensors summary : Counter({torch.Size([768]): 56, torch.Size([768, 768]): 24, torch.Size([3072, 768]): 6, torch.Size([3072]): 6, torch.Size([768, 3072]): 6, torch.Size([512, 768]): 2, torch.Size([512]): 1, torch.Size([119547, 768]): 1})
allocated tensors count 102
-----------------------
allocated tensors summary : Counter({torch.Size([768]): 56, torch.Size([768, 768]): 24, torch.Size([8, 8, 768]): 7, torch.Size([3072, 768]): 6, torch.Size([3072]): 6, torch.Size([768, 3072]): 6, torch.Size([512, 768]): 2, torch.Size([8, 512]): 2, torch.Size([8, 8]): 2, torch.Size([1, 8]): 2, torch.Size([512]): 1, torch.Size([119547, 768]): 1, torch.Size([8, 768]): 1})
allocated tensors count 116
-----------------------
allocated tensors summary : Counter({torch.Size([768]): 56, torch.Size([768, 768]): 24, torch.Size([8, 10, 768]): 7, torch.Size([3072, 768]): 6, torch.Size([3072]): 6, torch.Size([768, 3072]): 6, torch.Size([8, 512]): 3, torch.Size([512, 768]): 2, torch.Size([8, 10]): 2, torch.Size([1, 10]): 2, torch.Size([512]): 1, torch.Size([119547, 768]): 1, torch.Size([8, 768]): 1})
allocated tensors count 117
-----------------------
allocated tensors summary : Counter({torch.Size([768]): 56, torch.Size([768, 768]): 24, torch.Size([8, 9, 768]): 7, torch.Size([3072, 768]): 6, torch.Size([3072]): 6, torch.Size([768, 3072]): 6, torch.Size([8, 512]): 4, torch.Size([512, 768]): 2, torch.Size([8, 9]): 2, torch.Size([1, 9]): 2, torch.Size([512]): 1, torch.Size([119547, 768]): 1, torch.Size([8, 768]): 1})
allocated tensors count 118
-----------------------
allocated tensors summary : Counter({torch.Size([768]): 56, torch.Size([768, 768]): 24, torch.Size([8, 10, 768]): 7, torch.Size([3072, 768]): 6, torch.Size([3072]): 6, torch.Size([768, 3072]): 6, torch.Size([8, 512]): 5, torch.Size([512, 768]): 2, torch.Size([8, 10]): 2, torch.Size([1, 10]): 2, torch.Size([512]): 1, torch.Size([119547, 768]): 1, torch.Size([8, 9, 768]): 1, torch.Size([8, 768]): 1})
allocated tensors count 120
-----------------------
allocated tensors summary : Counter({torch.Size([768]): 56, torch.Size([768, 768]): 24, torch.Size([8, 9, 768]): 7, torch.Size([3072, 768]): 6, torch.Size([3072]): 6, torch.Size([768, 3072]): 6, torch.Size([8, 512]): 6, torch.Size([512, 768]): 2, torch.Size([8, 9]): 2, torch.Size([1, 9]): 2, torch.Size([512]): 1, torch.Size([119547, 768]): 1, torch.Size([8, 10, 768]): 1, torch.Size([8, 768]): 1})
allocated tensors count 121
-----------------------
allocated tensors summary : Counter({torch.Size([768]): 56, torch.Size([768, 768]): 24, torch.Size([8, 512]): 7, torch.Size([8, 10, 768]): 7, torch.Size([3072, 768]): 6, torch.Size([3072]): 6, torch.Size([768, 3072]): 6, torch.Size([512, 768]): 2, torch.Size([8, 10]): 2, torch.Size([1, 10]): 2, torch.Size([512]): 1, torch.Size([119547, 768]): 1, torch.Size([8, 9, 768]): 1, torch.Size([8, 768]): 1})
allocated tensors count 122
-----------------------
```
without np.copy() allocated tensors count increase +1 on every batch (this could be bad for the gpu memory if you are embedding a lot of samples at once)

whilst with ` embeddings = np.copy(embeddings.cpu().detach().numpy())` we observe the following trace during embeddings :+1: 
```
allocated tensors summary : Counter({torch.Size([768]): 56, torch.Size([768, 768]): 24, torch.Size([8, 9, 768]): 7, torch.Size([3072, 768]): 6, torch.Size([3072]): 6, torch.Size([768, 3072]): 6, torch.Size([512, 768]): 2, torch.Size([8, 9]): 2, torch.Size([1, 9]): 2, torch.Size([512]): 1, torch.Size([119547, 768]): 1, torch.Size([8, 10, 768]): 1, torch.Size([8, 768]): 1, torch.Size([8, 512]): 1})
allocated tensors count 116
-----------------------
allocated tensors summary : Counter({torch.Size([768]): 56, torch.Size([768, 768]): 24, torch.Size([8, 10, 768]): 7, torch.Size([3072, 768]): 6, torch.Size([3072]): 6, torch.Size([768, 3072]): 6, torch.Size([512, 768]): 2, torch.Size([8, 10]): 2, torch.Size([1, 10]): 2, torch.Size([512]): 1, torch.Size([119547, 768]): 1, torch.Size([8, 9, 768]): 1, torch.Size([8, 768]): 1, torch.Size([8, 512]): 1})
allocated tensors count 116
-----------------------
allocated tensors summary : Counter({torch.Size([768]): 56, torch.Size([768, 768]): 24, torch.Size([8, 11, 768]): 7, torch.Size([3072, 768]): 6, torch.Size([3072]): 6, torch.Size([768, 3072]): 6, torch.Size([512, 768]): 2, torch.Size([8, 11]): 2, torch.Size([1, 11]): 2, torch.Size([512]): 1, torch.Size([119547, 768]): 1, torch.Size([8, 10, 768]): 1, torch.Size([8, 768]): 1, torch.Size([8, 512]): 1})
allocated tensors count 116
-----------------------
allocated tensors summary : Counter({torch.Size([768]): 56, torch.Size([768, 768]): 24, torch.Size([8, 9, 768]): 7, torch.Size([3072, 768]): 6, torch.Size([3072]): 6, torch.Size([768, 3072]): 6, torch.Size([512, 768]): 2, torch.Size([8, 9]): 2, torch.Size([1, 9]): 2, torch.Size([512]): 1, torch.Size([119547, 768]): 1, torch.Size([8, 11, 768]): 1, torch.Size([8, 768]): 1, torch.Size([8, 512]): 1})
allocated tensors count 116
-----------------------
allocated tensors summary : Counter({torch.Size([768]): 56, torch.Size([768, 768]): 24, torch.Size([8, 9, 768]): 8, torch.Size([3072, 768]): 6, torch.Size([3072]): 6, torch.Size([768, 3072]): 6, torch.Size([512, 768]): 2, torch.Size([8, 9]): 2, torch.Size([1, 9]): 2, torch.Size([512]): 1, torch.Size([119547, 768]): 1, torch.Size([8, 768]): 1, torch.Size([8, 512]): 1})
allocated tensors count 116
-----------------------
allocated tensors summary : Counter({torch.Size([768]): 56, torch.Size([768, 768]): 24, torch.Size([8, 9, 768]): 8, torch.Size([3072, 768]): 6, torch.Size([3072]): 6, torch.Size([768, 3072]): 6, torch.Size([512, 768]): 2, torch.Size([8, 9]): 2, torch.Size([1, 9]): 2, torch.Size([512]): 1, torch.Size([119547, 768]): 1, torch.Size([8, 768]): 1, torch.Size([8, 512]): 1})
allocated tensors count 116
-----------------------

```
with extra np,copy() allocated tensors count remains stable (wich is good for the gpu memory :+1: ) 

my conclusions are :
- detaching tensors (even if they are casted as numpy array) is not enough for them to being garbage collected asap (some tensors references might be still hold by the torch session)
- so I forced natural python garbage collection with a clear separation between "tensors as numpy array" into pure numpy with the extra `np.copy()` operation

ps : i also tried the `torch.cuda.empty_cache()` but only the np.copy() does the trick.

**Comment by nreimers** [comment]
Thanks for your detailed explanations. I added the changed to the code and it will be released with the next version.
